File: hu/codes/python/chapter_array_and_linkedlist/array.py
"""
File: array.py
Created Time: 2022-11-25
Author: krahets (krahets@163.com)
"""

import random


def random_access(nums: list[int]) -> int:
    """Véletlenszerű hozzáférés elemhez"""
    # Véletlenszerűen választ egy számot a [0, len(nums)-1] intervallumból
    random_index = random.randint(0, len(nums) - 1)
    # A véletlenszerű elemet lekéri és visszaadja
    random_num = nums[random_index]
    return random_num


# Megjegyzés: a Python listája dinamikus tömb, közvetlenül bővíthető
# Tanulási célból ez a függvény a listát megváltoztathatatlan hosszúságú tömbként kezeli
def extend(nums: list[int], enlarge: int) -> list[int]:
    """Tömb hosszának bővítése"""
    # Bővített hosszúságú tömb inicializálása
    res = [0] * (len(nums) + enlarge)
    # Az eredeti tömb összes elemét az új tömbbe másolja
    for i in range(len(nums)):
        res[i] = nums[i]
    # A bővített új tömböt visszaadja
    return res


def insert(nums: list[int], num: int, index: int):
    """A num elemet az index pozícióba szúrja be a tömbben"""
    # Az index pozíciótól kezdve minden elemet egy hellyel hátrébb mozgat
    for i in range(len(nums) - 1, index, -1):
        nums[i] = nums[i - 1]
    # A num értéket az index pozícióba rendeli
    nums[index] = num


def remove(nums: list[int], index: int):
    """Az index pozícióban lévő elemet törli"""
    # Az index utáni összes elemet egy hellyel előrébb mozgat
    for i in range(index, len(nums) - 1):
        nums[i] = nums[i + 1]


def traverse(nums: list[int]):
    """Tömb bejárása"""
    count = 0
    # Tömb bejárása index szerint
    for i in range(len(nums)):
        count += nums[i]
    # Tömb elemeinek közvetlen bejárása
    for num in nums:
        count += num
    # Adat-index és elemek egyidejű bejárása
    for i, num in enumerate(nums):
        count += nums[i]
        count += num


def find(nums: list[int], target: int) -> int:
    """A megadott elem keresése a tömbben"""
    for i in range(len(nums)):
        if nums[i] == target:
            return i
    return -1


"""Fő kód"""
if __name__ == "__main__":
    # Tömb inicializálása
    arr = [0] * 5
    print("Tömb arr =", arr)
    nums = [1, 3, 2, 5, 4]
    print("Tömb nums =", nums)

    # Véletlenszerű hozzáférés
    random_num: int = random_access(nums)
    print("Véletlenszerű elem lekérése a nums-ból:", random_num)

    # Hossz bővítése
    nums: list[int] = extend(nums, 3)
    print("A tömb hosszának 8-ra bővítése után nums =", nums)

    # Elem beszúrása
    insert(nums, 6, 3)
    print("A 6-os szám 3-as indexre szúrása után nums =", nums)

    # Elem törlése
    remove(nums, 2)
    print("A 2-es indexű elem törlése után nums =", nums)

    # Tömb bejárása
    traverse(nums)

    # Elem keresése
    index: int = find(nums, 3)
    print("A 3-as elem keresése a nums-ban, index =", index)

File: hu/codes/java/chapter_array_and_linkedlist/array.java
/**
 * File: array.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_array_and_linkedlist;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class array {
    /* Véletlenszerű hozzáférés elemhez */
    static int randomAccess(int[] nums) {
        // Véletlenszerűen kiválaszt egy számot a [0, nums.length) intervallumból
        int randomIndex = ThreadLocalRandom.current().nextInt(0, nums.length);
        // Lekéri és visszaadja a véletlen elemet
        int randomNum = nums[randomIndex];
        return randomNum;
    }

    /* Tömb hosszának bővítése */
    static int[] extend(int[] nums, int enlarge) {
        // Inicializál egy bővített hosszú tömböt
        int[] res = new int[nums.length + enlarge];
        // Az eredeti tömb összes elemét átmásolja az új tömbbe
        for (int i = 0; i < nums.length; i++) {
            res[i] = nums[i];
        }
        // Visszaadja a bővített új tömböt
        return res;
    }

    /* num elem beszúrása az index pozícióba */
    static void insert(int[] nums, int num, int index) {
        // Az index pozíciótól kezdve minden elemet eggyel hátrébb lép
        for (int i = nums.length - 1; i > index; i--) {
            nums[i] = nums[i - 1];
        }
        // num-ot az index pozícióba rendeli
        nums[index] = num;
    }

    /* Az index pozícióban lévő elem törlése */
    static void remove(int[] nums, int index) {
        // Az index utáni összes elemet eggyel előrébb lép
        for (int i = index; i < nums.length - 1; i++) {
            nums[i] = nums[i + 1];
        }
    }

    /* Tömb bejárása */
    static void traverse(int[] nums) {
        int count = 0;
        // Tömb bejárása index szerint
        for (int i = 0; i < nums.length; i++) {
            count += nums[i];
        }
        // Tömb elemeinek közvetlen bejárása
        for (int num : nums) {
            count += num;
        }
    }

    /* A megadott elem megkeresése a tömbben */
    static int find(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == target)
                return i;
        }
        return -1;
    }

    /* Főprogram */
    public static void main(String[] args) {
        /* Tömb inicializálása */
        int[] arr = new int[5];
        System.out.println("Array arr = " + Arrays.toString(arr));
        int[] nums = { 1, 3, 2, 5, 4 };
        System.out.println("Array nums = " + Arrays.toString(nums));

        /* Elem beszúrása */
        int randomNum = randomAccess(nums);
        System.out.println("Get random element in nums " + randomNum);

        /* Tömb bejárása */
        nums = extend(nums, 3);
        System.out.println("Extend array length to 8, resulting in nums = " + Arrays.toString(nums));

        /* Elem beszúrása */
        insert(nums, 6, 3);
        System.out.println("Insert number 6 at index 3, resulting in nums = " + Arrays.toString(nums));

        /* Elem törlése */
        remove(nums, 2);
        System.out.println("Remove element at index 2, resulting in nums = " + Arrays.toString(nums));

        /* Tömb bejárása */
        traverse(nums);

        /* Elem keresése */
        int index = find(nums, 3);
        System.out.println("Find element 3 in nums, get index = " + index);
    }
}
